### ISSUE TITLE
Feature Proposal: Native Face & Object Recognition for Photos

### ISSUE THREAD

**Issue by paul43210** [owncloud/ocis#11916]
# Feature Proposal: Face & Object Recognition for oCIS Photos

**Author:** Paul Faure  
**Date:** January 18, 2026  
**Type:** Feature Proposal / Architecture Discussion

---

## Summary

I'd like to propose adding face detection, recognition, and object tagging capabilities to oCIS, enabling users to search their photos by people ("find all photos with Mom") and content ("show me beach photos"). This builds on the existing Tika integration and Bleve search infrastructure.

Before investing significant development effort, I wanted to gauge the oCIS team's interest and identify any architectural concerns or philosophical objections early.

---

### User Expectations Have Changed

Photo management is increasingly important for personal and family cloud users. Services like Google Photos and Apple Photos have set expectations around:

- Automatic face detection and grouping
- Searching photos by person name
- Object/scene recognition ("beach", "dog", "birthday party")

### The Current Landscape

Several open-source solutions offer photo intelligence today:

| Solution | Approach | Limitations |
|----------|----------|-------------|
| **PhotoPrism** | Standalone photo manager with excellent AI | Separate UI, separate storage, complex sync with cloud storage |
| **Immich** | Self-hosted Google Photos alternative | Designed as complete replacement, not integration layer |
| **Nextcloud + Recognize** | App-based face/object recognition | Tied to Nextcloud ecosystem |

**The common problem:** All of these require either replacing your file sync solution entirely or running parallel systems with clunky integration. Users end up with:
- Split UIs (files in one place, photos in another)
- Duplicate storage or complex mount configurations
- Authentication silos
- Inconsistent sharing models

### Why oCIS-Native Matters

I chose oCIS for my family cloud specifically because of its robust, reliable, and performant architecture. The microservices design, Bleve search, and clean API are excellent foundations. However, when it comes to photo management, I'm forced to consider:

1. **Run PhotoPrism alongside oCIS** → Two UIs, storage sync headaches, separate user management
2. **Switch to Immich entirely** → Lose oCIS's superior file sync, collaboration, and office integration
3. **Accept limited photo features** → Miss out on face/object search that families now expect

None of these are satisfying. What I'd prefer is photo intelligence that's **native to oCIS** - same UI, same storage, same sharing model, same user accounts.

### The Opportunity

oCIS already has most of the infrastructure needed:
- **Tika integration** for content extraction (including `ObjectRecognitionParser` for ML services)
- **Bleve search** with custom field support
- **NATS event bus** for triggering processing on upload
- **Web extension architecture** for custom UI
- **Graph/WebDAV APIs** for metadata

Adding face and object recognition would:
1. **Close a significant feature gap** for family/personal use cases
2. **Leverage existing infrastructure** rather than bolting on external tools
3. **Maintain privacy** with on-premises processing (no cloud AI services)
4. **Preserve oCIS's unified experience** - one UI, one storage, one sharing model
5. **Differentiate oCIS** from solutions that require PhotoPrism/Immich integration

---

### High-Level Overview

```
┌─────────────────────────────────────────────────────────────────────────┐
│                              oCIS Server                                 │
│                                                                          │
│  ┌────────────────────────────────────────────────────────────────────┐ │
│  │                    Search Service (extended)                        │ │
│  │  - Existing: content extraction via Tika                           │ │
│  │  - New: parse face embeddings + objects from Tika response         │ │
│  │  - New: call Face Service for clustering                           │ │
│  │  - New: index face names + objects in Bleve                        │ │
│  └─────────────────────────────────┬──────────────────────────────────┘ │
│                                    │                                     │
│  ┌─────────────────────────────────▼──────────────────────────────────┐ │
│  │                    Face Service (NEW)                               │ │
│  │                                                                     │ │
│  │  - Per-user face cluster storage (SQLite)                          │ │
│  │  - Embedding similarity matching (cosine distance)                 │ │
│  │  - Cluster naming API for web UI                                   │ │
│  │  - Complete user isolation (User A's names never visible to B)     │ │
│  └─────────────────────────────────────────────────────────────────────┘ │
│                                                                          │
│  ┌─────────────────────────────────────────────────────────────────────┐ │
│  │                    Web Extension: face-manager                      │ │
│  │  - "Unknown Faces" view for naming detected faces                  │ │
│  │  - "People" view for browsing by person                            │ │
│  │  - Face overlays in photo viewer                                   │ │
│  └─────────────────────────────────────────────────────────────────────┘ │
└──────────────────────────────────────────────────────────────────────────┘
                                    │
                           HTTP (existing Tika flow)
                                    │
                                    ▼
┌──────────────────────────────────────────────────────────────────────────┐
│                         Tika Server (extended config)                     │
│                                                                           │
│  ObjectRecognitionParser configured to call:                             │
│  - Face detection service (InsightFace) → returns embeddings            │
│  - Object recognition service (Inception) → returns labels              │
│                                                                           │
│  Returns to oCIS:                                                        │
│  {                                                                        │
│    "content": "...",                                                     │
│    "faces": [{ "bbox": [...], "embedding": [512 floats] }],             │
│    "objects": ["beach", "sunset", "dog"]                                 │
│  }                                                                        │
└──────────────────────────────────────────────────────────────────────────┘
                                    │
                    ┌───────────────┴───────────────┐
                    ▼                               ▼
        ┌───────────────────────┐     ┌─────────────────────────────┐
        │  InsightFace-REST     │     │  Inception/Object Service   │
        │  (stateless)          │     │  (stateless)                │
        │                       │     │                             │
        │  Face detection +     │     │  ~1000 object categories    │
        │  512-dim embeddings   │     │  "beach", "car", "dog"...   │
        │                       │     │                             │
        │  No user knowledge    │     │  No user knowledge          │
        └───────────────────────┘     └─────────────────────────────┘
```

### Why Not Just Integrate PhotoPrism?

I explored deep PhotoPrism integration before designing this proposal. The challenges:

| Challenge | Impact |
|-----------|--------|
| **Single-user architecture** | PhotoPrism doesn't support multi-tenant face isolation; one global face database |
| **Storage model mismatch** | PhotoPrism wants to manage originals; sharing oCIS's decomposed filesystem is discouraged |
| **Separate authentication** | No native OIDC integration with oCIS's IDP |
| **UI fragmentation** | Users must switch between oCIS web and PhotoPrism for different tasks |
| **Sharing model incompatibility** | PhotoPrism shares work differently than oCIS spaces/links |

A native oCIS solution avoids all of these by keeping everything within oCIS's existing patterns.

### Key Design Principles

1. **Stateless ML backends** - InsightFace and object recognition services know nothing about users, names, or oCIS. They simply process images and return embeddings/labels.

2. **oCIS owns the identity mapping** - The Face Service within oCIS manages the mapping from face embeddings to user-assigned names. This keeps sensitive data (who is in photos) within oCIS's security boundary.

3. **Per-user isolation** - Each user has their own face database. If Paul names a face "Mom" and Sheila names the same face "Aunt Jane", both are valid in their respective namespaces. No cross-user data leakage.

4. **Builds on existing infrastructure**:
   - Tika's `ObjectRecognitionParser` already supports calling external REST services
   - Bleve already indexes custom fields (demonstrated with photo EXIF metadata)
   - NATS events already trigger content extraction on upload
   - Web extensions architecture already supports custom views

5. **Optional deployment** - Face/object recognition would be opt-in, similar to how Tika content extraction is optional today. Users who don't want it don't pay the resource cost.

---

### Naming Unknown Faces

When faces are detected, they're grouped by visual similarity into clusters. Users see an "Unknown Faces" view where they can name each cluster:

```
┌─────────────────────────────────────────────────────────────────┐
│  Unknown Faces (23 to name)                                     │
│                                                                 │
│  ┌─────────────┐  ┌─────────────┐  ┌─────────────┐             │
│  │ [stack of   │  │             │  │ [stack of   │             │
│  │  47 similar │  │  [1 photo]  │  │  12 similar │             │
│  │  faces]     │  │             │  │  faces]     │             │
│  │             │  │             │  │             │             │
│  │ [Select ▼]  │  │ [Select ▼]  │  │ [Select ▼]  │             │
│  │  Mom        │  │             │  │             │             │
│  │  Dad        │  │             │  │             │             │
│  │  + New...   │  │             │  │             │             │
│  └─────────────┘  └─────────────┘  └─────────────┘             │
└─────────────────────────────────────────────────────────────────┘
```

Once named, future photos with that person are automatically tagged.

### Search Integration

Fully integrated with existing oCIS search:

```
face:Mom                           → All photos with Mom
face:Mom AND face:Dad              → Photos with both parents
object:beach AND face:*            → Beach photos with people
object:dog                         → All photos containing dogs
has_unnamed_faces:true             → Photos needing face tagging
content:"birthday" AND face:John   → Birthday documents/photos with John
```

### Unified Experience

Unlike PhotoPrism/Immich integration approaches:
- **One UI** - Face management lives in oCIS web, not a separate app
- **One storage** - No sync, no mounts, no duplicate files
- **One sharing model** - Share a photo with faces, recipient sees the names
- **One search** - Query faces, objects, content, EXIF, and tags together
- **One mobile app** - When oCIS mobile improves, photo features come along

---

## Sharing Behavior

When a photo is shared:
- Face names travel with the photo as **read-only metadata**
- Recipient sees labels but cannot edit them
- This does NOT train the recipient's face recognition
- Recipient's own uploads are processed independently in their namespace

This maintains privacy while allowing meaningful sharing - exactly how oCIS sharing should work.

---

## Implementation Scope

| Component | Estimated Effort | Notes |
|-----------|------------------|-------|
| Face Service (Go) | ~1500 lines | Clustering, storage, API |
| Search Service mods | ~400 lines | Parse Tika output, Bleve indexing |
| Graph API extensions | ~300 lines | REST endpoints |
| Web Extension (Vue.js) | ~1800 lines | UI for naming, browsing |
| Tika configuration | Minimal | XML config for ObjectRecognitionParser |

**Total:** ~4000 lines of new code, plus deployment of InsightFace and object recognition containers (existing open-source Docker images).

---

## Questions for the oCIS Team

1. **Alignment with oCIS direction** - Is photo intelligence (face/object recognition) something oCIS wants to support natively, or is integration with external tools like PhotoPrism the preferred path?

2. **Privacy/security concerns** - Any concerns with storing face embeddings (biometric-adjacent data) in oCIS? The embeddings cannot reconstruct faces, but they can identify if two photos contain the same person.

3. **Architecture feedback** - Does the proposed architecture (Face Service as new oCIS service, Tika as ML orchestrator) align with oCIS patterns? Or would a different approach be preferred?

4. **ML backend preferences** - Any preference on face detection technology? InsightFace (open source, state-of-the-art accuracy, MIT license) vs alternatives?

5. **Contribution path** - If there's interest, would this be considered for upstream contribution, or better maintained as a community extension?

6. **Existing work** - Is anyone else working on similar functionality? Happy to collaborate rather than duplicate effort.

---

## About Me

I'm running a custom oCIS deployment for my family (~20GB of photos across multiple users). I've built a [photo-addon extension](https://github.com/paul43210/ocis-photo-addon) that provides timeline views with date grouping and EXIF-based browsing. I recently contributed [PR #11912](https://github.com/owncloud/ocis/pull/11912) adding photo metadata fields (camera make/model, GPS coordinates, capture date) to the Bleve search index.

I have free cycles on my Claude Code and looking to put them towards good use. Yes AI does all my heavy lifting!

I evaluated PhotoPrism, Immich, and Nextcloud Recognize before concluding that oCIS-native photo intelligence would be the cleanest solution. I'm happy to develop this independently if it's outside oCIS's scope, but wanted to check alignment before building something that might conflict with the project's direction - or worse, duplicate work someone else is already doing.

Looking forward to your feedback!

---

## References

- [InsightFace](https://github.com/deepinsight/insightface) - State-of-the-art face detection/recognition (MIT license, 99.86% LFW accuracy)
- [Tika ObjectRecognitionParser](https://cwiki.apache.org/confluence/display/tika/TikaAndVision) - Existing Tika support for calling ML services
- [PhotoPrism](https://photoprism.app/) - Standalone photo manager (comparison reference)
- [Immich](https://immich.app/) - Self-hosted Google Photos alternative (comparison reference)
- [My photo-addon extension](https://github.com/paul43210/ocis-photo-addon) - Existing work on oCIS photo browsing
- [PR #11912](https://github.com/owncloud/ocis/pull/11912) - My contribution adding photo EXIF fields to Bleve

**Comment by jvillafanez** [comment]
Just my personal worries:

> Automatic face detection and grouping

Why is this needed, specially in a personal / family environment? It seems a cool toy with little purpose.

> Searching photos by person name

Technically, you could do it "now" by indexing the appropriate data. I'd rather include custom metadata if possible.
If you want to include "people" as custom metadata, you could; if someone in a university have photos or drawings of different nuclear reactors and wants to catalog them, he could.

We'd probably need a custom web UI to allow the users to include this metadata, and have some control over who's adding it, but I think most of the stuff is already there. This seems more flexible, and the maintenance cost shouldn't skyrocket.

> Object/scene recognition ("beach", "dog", "birthday party")

This is bound to make mistakes. Even if there is a 99% of success, it's the 1% of failure what makes it unreliable. This means that someone will need to manually check the validity of the recognition. At that point, it seems better to tag the files and index them the way you want.

Tags should work fine for this use case, with a 100% success rate unless you're tagging things wrong on purpose.

-----

My main worry is the maintenance cost. Most of the oCIS stuff is or can be offloaded because there are already 3rd party solutions providing the feature you need in a better way, with barely any maintenance cost (other than the integration itself and the software upgrades).

I'd personally go to include custom searchable fields. Ideally, it should be just providing the custom fields in the UI and index them in the search engine. The overall cost should be cut by half easily, it's a more flexible solution and everyone can benefit from it.

**Comment by mklos-kw** [comment]
I love the idea and the functionality. Indeed vector search can provide non-100% recall than explicit tags but deeper semantic understanding is good alternative to keyword or full-text search alone.

I agree on the maintenance cost concerns. Given that, I'd suggest considering an MVP for vector/semantic search first. Face intelligence could be an experimental layer on top.

How about UX similar to **Pinterest Lens** or **Google Photos "More Like This"** - users upload a reference image and get visually/semantically similar results from their own library.

**Data flow:**
1. Image upload triggers existing Tika content extraction flow
2. Tika's ObjectRecognitionParser calls CPU inference service → returns embedding vector
3. Tika response includes embedding: `{ "content": "...", "embedding": [N floats] }`
4. Search Service parses embedding from Tika response and indexes it in vector search
5. Users search by reference image (upload or select existing) → read/compute embedding → vector search returns ranked matches

**Proposed Minimal Architecture:**
- **Search Service (extended)** - as mentioned
- **Tika Server (extended config)** as mentioned
- **CPU inference service** - runs image embedding models (can consume TensorFlow or PyTorch models, ideally portable for less friction). 

Notable embeddable go libraries:

**Vector indexing/search:**
- [kelindar/search](https://github.com/kelindar/search) (515 stars) - Pure Go, SIMD-optimized brute-force search, file-based storage, no CGO. Best for embedded use with moderate dataset sizes (~8ns dot product operations).
- [blevesearch/go-faiss](https://github.com/blevesearch/go-faiss) (9 stars) - FAISS CGO bindings, natural fit since oCIS uses Bleve, requires FAISS C library compilation.

**CPU inference:**
- [yalue/onnxruntime_go](https://github.com/yalue/onnxruntime_go) (540 stars) - ONNX Runtime CGO bindings, actively maintained (last update Jan 2026), supports CPU/GPU execution providers. PyTorch/TensorFlow models can be exported to ONNX format.
- [mattn/go-tflite](https://github.com/mattn/go-tflite) (313 stars) - TensorFlow Lite CGO bindings, actively maintained (last release May 2024), optimized for mobile/edge CPU inference.
- [sugarme/gotch](https://github.com/sugarme/gotch) (636 stars) - PyTorch LibTorch CGO bindings, comprehensive API (2525+ tensor operations), supports JIT models and pretrained model loading.


I hope this will help you in more stability in photo management experiments on images one can on their own drive :)
